internal/sdkv2provider/resource_cloudflare_zone_cache_reserve.go

```diff
@@ -0,0 +1,138 @@
+package sdkv2provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/MakeNowJust/heredoc/v2"
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/cloudflare/terraform-provider-cloudflare/internal/consts"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	cacheReserveEnabled  = "on"
+	cacheReserveDisabled = "off"
+)
+
+func resourceCloudflareZoneCacheReserve() *schema.Resource {
+	return &schema.Resource{
+		Schema:        resourceCloudflareZoneCacheReserveSchema(),
+		CreateContext: resourceCloudflareZoneCacheReserveCreate,
+		ReadContext:   resourceCloudflareZoneCacheReserveRead,
+		UpdateContext: resourceCloudflareZoneCacheReserveUpdate,
+		DeleteContext: resourceCloudflareZoneCacheReserveDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceCloudflareZoneCacheReserveImport,
+		},
+		Description: heredoc.Doc(`
+			Provides a Cloudflare Cache Reserve resource. Cache Reserve can
+			increase cache lifetimes by automatically storing all cacheable
+			files in Cloudflare's persistent object storage buckets.
+
+			Note: Using Cache Reserve without Tiered Cache is not recommended.
+
+			Requires Cache Reserve subscription.
+		`),
+	}
+}
+
+func resourceCloudflareZoneCacheReserveCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	zoneID := d.Get(consts.ZoneIDSchemaKey).(string)
+
+	// Ensure an unique resource ID to differentiate it from a Zone ID.
+	d.SetId(stringChecksum(fmt.Sprintf("%s/cache-reserve", zoneID)))
+
+	return resourceCloudflareZoneCacheReserveUpdate(ctx, d, meta)
+}
+
+func resourceCloudflareZoneCacheReserveRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get(consts.ZoneIDSchemaKey).(string)
+
+	tflog.Info(ctx, "reading Cache Reserve", map[string]interface{}{
+		"zone_id": zoneID,
+	})
+
+	params := cloudflare.GetCacheReserveParams{}
+	output, err := client.GetCacheReserve(ctx, cloudflare.ZoneIdentifier(zoneID), params)
+	if err != nil {
+		var notFoundError *cloudflare.NotFoundError
+		if errors.As(err, &notFoundError) {
```

Comment:
@jacobbednarz, would this ever be a problem for Cache Reserve?

While testing the code during development, when I selected a non-existent zone ID, albeit still a valid format, I got a permissions/access error rather than a 404. I am unsure how this is implemented on the backend side, though.

That said, I didn't check a non-sensical ID such as "this is a test" to something along these lines. I will do it later to see what will happen.

Reply:
I will answer my question: **GetCacheReserve()** returns a **404**, while **UpdateCacheReserve()** returns a different type of error message. I also ensured that 404 is handled correctly in the **Delete** callback.